ZOOKEEPER-1782: Let a SASL super user be super

revans2 opened the pull request.

hanm commented on src/java/test/org/apache/zookeeper/test/SaslSuperUserTest.java:
```diff
@@ -0,0 +1,118 @@
+        }
+    }
+
+    private AtomicInteger authFailed = new AtomicInteger(0);
```
Do we need this? Seems it's not used anywhere.

hanm commented on src/java/test/org/apache/zookeeper/test/SaslSuperUserTest.java:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    @Test
+    public void testSuperIsSuper() throws Exception {
```
This test does not verify anything. Maybe put an assert in a catch block to indicate no auth error should happen during these operations?

hanm commented on src/java/test/org/apache/zookeeper/test/SaslSuperUserTest.java:
```diff
@@ -0,0 +1,118 @@
+            zk.delete("/digest_read", -1);
+            zk.delete("/sasl_read/sub", -1);
+            zk.delete("/sasl_read", -1);
+            Thread.sleep(1000);
```
Can this sleep be removed?

revans2 commented:
@hanm Thanks for the review.  I think I have addressed all of your review comments.  I originally put the patch up in 2013 and I have not touched it since so I honestly don't remember a lot about why I made the test that way to begin with.

arshadmohammad commented on src/java/main/org/apache/zookeeper/server/auth/SASLAuthenticationProvider.java:
```diff
@@ -38,11 +38,6 @@ public String getScheme() {
-              return true;
-            }
-        }
         if ((id.equals("super") || id.equals(aclExpr))) {
```
I think we should remove (id.equals("super") also. Lets have only one way being super user.

revans2 replied:
Sure.  It looks like I am going to have to pull out the "super" user SASL Login too.

https://github.com/apache/zookeeper/blob/master/src/java/main/org/apache/zookeeper/server/auth/SaslServerCallbackHandler.java#L42
https://github.com/apache/zookeeper/blob/master/src/java/main/org/apache/zookeeper/server/auth/SaslServerCallbackHandler.java#L99-L101

Do you know where the source for the documentation at https://zookeeper.apache.org/doc/r3.3.3/zookeeperAdmin.html is? Under the section "Authentication & Authorization Options" it talks about how to use the "super" user. 

I also found docs for this in a few other places on the open internet and because this is a backwards incompatible change I get a little nervous, so is there anything else I need to do to document a breaking change like this?

arshadmohammad replied:
Thanks @revans2 for the details.
On second thought, it would not be appropriate to break the backward compatibility. So can you please revert changes done for this comment. Rest all looks good to me.

revans2 replied:
done

arshadmohammad commented on src/java/test/org/apache/zookeeper/test/SaslSuperUserTest.java:
```diff
@@ -0,0 +1,118 @@
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.ZooKeeper;
+import org.apache.zookeeper.Watcher.Event.KeeperState;
+import org.apache.zookeeper.ZooDefs.Ids;
```
The import org.apache.zookeeper.ZooDefs.Ids is never used.
There are many other unused imports as well

arshadmohammad commented on src/java/test/org/apache/zookeeper/test/SaslSuperUserTest.java:
```diff
@@ -0,0 +1,118 @@
+    private static Id otherDigestUser;
+ 
+    static {
+        System.setProperty("zookeeper.authProvider.1","org.apache.zookeeper.server.auth.SASLAuthenticationProvider");
```
Whatever the system properties are set in this class should be cleared at the end of the class. 
Similar test class  org.apache.zookeeper.SaslAuthTest could be referred.

revans2 commented:
@arshadmohammad I think I have addressed your review comments

revans2 commented:
@arshadmohammad if everything looks good I will squash my commits

hanm commented:
One more thing here. Looks like we don't have an existing configuration option or system property called "superUser" (or zookeeper.superUser). So this `zookeeper.superUser` should be documented.

The existing approach of how superuser is defined under various auth provider:
* For digest, it's `zookeeper.DigestAuthenticationProvider.superDigest`.
* For ssl, it's `zookeeper.X509AuthenticationProvider.superUser`.

So maybe for SASL we name it `zookeeper.SASLAuthenticationProvider.superUser`, rather than `zookeeper.superUser` given it's only applicable to SASL?

revans2 commented:
@hanm I didn't add in `zookeeper.superUser`.  I am happy to document it if you tell me where to make that change.  I don't know where the documentation source is stored.  I am happy to add in an additional `zookeeper.SASLAuthenticationProvider.superUser` config and document it instead, but removing `zookeeper.superUser` will technically break backwards compatibility.  If that is what you want I can do it, I just want to be sure like on the other breaking change that was suggested.

hanm commented:
`zookeeper.superUser` was added in ZOOKEEPER-938 and it was never documented. I doubt people even know about this property without looking into source code and/or the description of ZOOKEEPER-938, but I agree compatibility concern here is very reasonable. 

So let's stick to `zookeeper.superUser` for this patch. We can give it a shinny name later. The only remaining work is to document this `zookeeper.superUser`. The doc source is located https://github.com/apache/zookeeper/blob/master/src/docs/src/documentation/content/xdocs/zookeeperAdmin.xml, search `zookeeper.DigestAuthenticationProvider.superDigest` or `X509AuthenticationProvider.superUser` as references of existing superuser documents.

revans2 commented:
@hanm I added in the docs.

hanm commented:
merged, thanks for the work @revans2 !
